File: totoggle_go/internal/strategy/matchlist_test.go
package strategy

import (
	"testing"

	"github.com/stretchr/testify/assert"

	"github.com/manorfm/toToggles/totoggle_go/internal/toggle"
)

// MatchListEvaluator is the one implementation shared by parameter, user_id, country, and canary
// — all four are "comma-separated allowlist, exact trimmed match" per the confirmed prototype
// hints, so this is a single reused type rather than four near-identical copies.
func TestMatchListEvaluator_MatchesOneOfCommaSeparatedValues(t *testing.T) {
	e := MatchListEvaluator{}
	rule := toggle.ActivationRule{Type: toggle.RuleTypeParameter, Value: "premium,enterprise"}

	assert.True(t, e.Evaluate(rule, "premium", true))
	assert.True(t, e.Evaluate(rule, "enterprise", true))
	assert.False(t, e.Evaluate(rule, "basic", true))
}

func TestMatchListEvaluator_TrimsWhitespaceAroundEntries(t *testing.T) {
	e := MatchListEvaluator{}
	rule := toggle.ActivationRule{Type: toggle.RuleTypeCountry, Value: " BR , US ,CA"}

	assert.True(t, e.Evaluate(rule, "BR", true))
	assert.True(t, e.Evaluate(rule, "US", true))
	assert.True(t, e.Evaluate(rule, "CA", true))
	assert.False(t, e.Evaluate(rule, "br", true))
}

func TestMatchListEvaluator_NoKeyNeverMatches(t *testing.T) {
	e := MatchListEvaluator{}
	rule := toggle.ActivationRule{Type: toggle.RuleTypeUserID, Value: "user-1,user-2"}

	assert.False(t, e.Evaluate(rule, "", false))
}

func TestMatchListEvaluator_BlankRuleValueNeverMatches(t *testing.T) {
	e := MatchListEvaluator{}
	rule := toggle.ActivationRule{Type: toggle.RuleTypeCanary, Value: "   "}

	assert.False(t, e.Evaluate(rule, "true", true))
}

func TestMatchListEvaluator_EmptyStringKeyCanMatchAnEmptyListEntry(t *testing.T) {
	e := MatchListEvaluator{}
	rule := toggle.ActivationRule{Type: toggle.RuleTypeCanary, Value: "true,"}

	assert.True(t, e.Evaluate(rule, "", true))
}

File: totoggle_node/src/internal/strategy/matchlist.test.ts
import { describe, expect, it } from "vitest";
import type { ActivationRule } from "../toggle/rule.js";
import { MatchListEvaluator } from "./matchlist.js";

// MatchListEvaluator is the one implementation shared by parameter, user_id, country, and canary
// — all four are "comma-separated allowlist, exact trimmed match" per the confirmed prototype
// hints, so this is a single reused type rather than four near-identical copies.
describe("MatchListEvaluator", () => {
  const evaluator = new MatchListEvaluator();

  it("matches one of the comma-separated values", () => {
    const rule: ActivationRule = { type: "parameter", value: "premium,enterprise" };
    expect(evaluator.evaluate(rule, "premium")).toBe(true);
    expect(evaluator.evaluate(rule, "enterprise")).toBe(true);
    expect(evaluator.evaluate(rule, "basic")).toBe(false);
  });

  it("trims whitespace around entries", () => {
    const rule: ActivationRule = { type: "country", value: " BR , US ,CA" };
    expect(evaluator.evaluate(rule, "BR")).toBe(true);
    expect(evaluator.evaluate(rule, "US")).toBe(true);
    expect(evaluator.evaluate(rule, "CA")).toBe(true);
    expect(evaluator.evaluate(rule, "br")).toBe(false);
  });

  it("never matches with no key at all", () => {
    const rule: ActivationRule = { type: "user_id", value: "user-1,user-2" };
    expect(evaluator.evaluate(rule, undefined)).toBe(false);
  });

  it("never matches a blank rule value", () => {
    const rule: ActivationRule = { type: "canary", value: "   " };
    expect(evaluator.evaluate(rule, "true")).toBe(false);
  });

  it("an empty-string key can match an empty list entry", () => {
    const rule: ActivationRule = { type: "canary", value: "true," };
    expect(evaluator.evaluate(rule, "")).toBe(true);
  });
});
